File: scripts/scan_secrets.py
import argparse
import re
import subprocess
import sys
from pathlib import Path


# 只做本地粗扫，目标是尽快发现明显的误提交风险。
ROOT = Path(__file__).resolve().parents[1]
DEFAULT_EXCLUDE_DIRS = {
    ".git",
    ".idea",
    ".pytest_cache",
    ".mypy_cache",
    ".ruff_cache",
    ".venv",
    "venv",
    "env",
    "__pycache__",
    "build",
    "dist",
}
DEFAULT_EXCLUDE_SUFFIXES = {
    ".png",
    ".jpg",
    ".jpeg",
    ".gif",
    ".webp",
    ".ico",
    ".pdf",
    ".zip",
    ".7z",
    ".rar",
    ".exe",
    ".dll",
    ".pyd",
    ".so",
    ".wav",
    ".mp3",
    ".mp4",
}
PLACEHOLDER_VALUES = {
    "",
    "null",
    "none",
    "changeme",
    "your-key",
    "your_api_key",
    "your-api-key",
    "your-access-key",
    "your-secret-key",
    "your token",
    "your access key",
    "your secret key",
    "your app key",
    "你的token",
    "你的密钥",
    "你的 api key",
    "你的 access key",
    "你的 app key",
    "你的阿里云百炼 api key",
    "你的豆包 access key",
    "你的豆包 app key",
}

SECRET_PATTERNS = [
    ("OpenAI-style key", re.compile(r"sk-[A-Za-z0-9_-]{16,}")),
    ("GitHub PAT", re.compile(r"github_pat_[A-Za-z0-9_]{20,}")),
    ("GitHub token", re.compile(r"ghp_[A-Za-z0-9]{20,}")),
    ("Slack token", re.compile(r"xox[baprs]-[A-Za-z0-9-]{10,}")),
    ("AWS access key", re.compile(r"AKIA[0-9A-Z]{16}")),
    (
        "Private key block",
        re.compile(r"-----BEGIN (?:RSA|DSA|EC|OPENSSH|PGP) PRIVATE KEY-----"),
    ),
]

ASSIGNMENT_PATTERN = re.compile(
    r"""(?ix)
    (?P<key>
        api[_-]?key|
        access[_-]?key|
        app[_-]?key|
        secret|
        secret[_-]?key|
        client[_-]?secret|
        token|
        password|
        passwd|
        pwd
    )
    \s*
    (?:
        [:=]
        |
        => 
    )
    \s*
    (?:
        (?P<q1>["'])(?P<quoted_value>.*?)(?P=q1)
        |
        (?P<bare_value>[A-Za-z0-9_./:+-]{8,})
    )
    \s*$
    """
)


def is_probably_binary(path: Path) -> bool:
    if path.suffix.lower() in DEFAULT_EXCLUDE_SUFFIXES:
        return True
    try:
        chunk = path.read_bytes()[:2048]
    except OSError:
        return True
    return b"\x00" in chunk


def should_skip(path: Path) -> bool:
    parts = set(path.parts)
    if parts & DEFAULT_EXCLUDE_DIRS:
        return True
    return is_probably_binary(path)


def looks_like_placeholder(value: str) -> bool:
    normalized = value.strip().strip('"').strip("'").lower()
    if not normalized:
        return True
    if normalized in PLACEHOLDER_VALUES:
        return True
    if normalized.isdigit():
        return True
    if normalized in {"true", "false"}:
        return True
    return any(word in normalized for word in ("example", "placeholder", "示例", "填写", "这里填"))


def scan_file(path: Path) -> list[tuple[int, str, str]]:
    findings: list[tuple[int, str, str]] = []
    try:
        text = path.read_text(encoding="utf-8")
    except UnicodeDecodeError:
        text = path.read_text(encoding="utf-8", errors="ignore")
    except OSError:
        return findings

    for line_no, line in enumerate(text.splitlines(), start=1):
        stripped = line.strip()
        if not stripped or stripped.startswith(("#", "//")):
            continue
        if stripped.count("|") >= 2:
            continue

        for name, pattern in SECRET_PATTERNS:
            match = pattern.search(line)
            if match:
                findings.append((line_no, name, mask_secret(match.group(0))))

        assignment = ASSIGNMENT_PATTERN.search(stripped)
        if assignment:
            value = assignment.group("quoted_value") or assignment.group("bare_value") or ""
            value = value.strip().strip('"').strip("'")
            if value and len(value) >= 8 and not looks_like_placeholder(value):
                if re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", value):
                    continue
                findings.append((line_no, f"Suspicious {assignment.group('key')}", mask_secret(value)))

    return findings


def mask_secret(value: str) -> str:
    if len(value) <= 8:
        return "*" * len(value)
    return f"{value[:4]}...{value[-4:]}"


def iter_files(root: Path):
    for path in root.rglob("*"):
        if not path.is_file():
            continue
        if should_skip(path.relative_to(root)):
            continue
        yield path


def get_git_root() -> Path | None:
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            capture_output=True,
            text=True,
            check=True,
        )
    except (OSError, subprocess.CalledProcessError):
        return None
    return Path(result.stdout.strip())


def get_staged_files(repo_root: Path) -> list[Path]:
    try:
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only", "--diff-filter=ACMR"],
            cwd=repo_root,
            capture_output=True,
            text=True,
            check=True,
        )
    except (OSError, subprocess.CalledProcessError):
        return []

    files: list[Path] = []
    for line in result.stdout.splitlines():
        relative = line.strip()
        if not relative:
            continue
        path = repo_root / relative
        if path.is_file():
            files.append(path)
    return files


def main() -> int:
    parser = argparse.ArgumentParser(description="Scan repository for likely secrets.")
    parser.add_argument("path", nargs="?", default=".", help="Directory to scan")
    parser.add_argument(
        "--staged",
        action="store_true",
        help="Scan only files staged for commit",
    )
    args = parser.parse_args()

    if args.staged:
        repo_root = get_git_root()
        if repo_root is None:
            print("[scan-secrets] not inside a git repository")
            return 2
        target = repo_root
        files = get_staged_files(repo_root)
        if not files:
            print("[scan-secrets] no staged files to scan")
            return 0
    else:
        target = Path(args.path).resolve()
        if not target.exists():
            print(f"[scan-secrets] path not found: {target}")
            return 2
        files = [target] if target.is_file() else list(iter_files(target))

    findings_count = 0

    for file_path in files:
        if not args.staged and file_path.is_file() and target.is_file():
            rel = file_path
            if should_skip(Path(file_path.name)):
                continue
        else:
            rel = file_path.relative_to(target)

        findings = scan_file(file_path)
        for line_no, kind, masked in findings:
            findings_count += 1
            print(f"{rel}:{line_no}: {kind}: {masked}")

    if findings_count:
        print(f"[scan-secrets] found {findings_count} potential secret(s)")
        return 1

    print("[scan-secrets] no obvious secrets found")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/scan-secrets.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";

const EXCLUDE_DIRS = new Set([
  ".git",
  ".idea",
  ".svelte-kit",
  ".venv",
  "build",
  "dist",
  "env",
  "node_modules",
  "package",
  "target",
  "venv",
]);

const BINARY_SUFFIXES = new Set([
  ".7z",
  ".dll",
  ".exe",
  ".gif",
  ".ico",
  ".icns",
  ".jpg",
  ".jpeg",
  ".mp3",
  ".mp4",
  ".pdf",
  ".png",
  ".pyd",
  ".rar",
  ".so",
  ".wav",
  ".webp",
  ".zip",
]);

const PROTECTED_LOCAL_FILES = new Set([
  "config.toml",
  "config.local.toml",
  "voice_input.log",
  "voice_input_stats.jsonl",
]);

const PLACEHOLDER_VALUES = new Set([
  "",
  "changeme",
  "null",
  "none",
  "your key",
  "your-key",
  "your_api_key",
  "your-api-key",
  "your-access-key",
  "your-secret-key",
  "your token",
  "your access key",
  "your secret key",
  "your app key",
]);

const SECRET_PATTERNS = [
  ["OpenAI-style key", /sk-[A-Za-z0-9_-]{16,}/],
  ["GitHub PAT", /github_pat_[A-Za-z0-9_]{20,}/],
  ["GitHub token", /ghp_[A-Za-z0-9]{20,}/],
  ["Slack token", /xox[baprs]-[A-Za-z0-9-]{10,}/],
  ["AWS access key", /AKIA[0-9A-Z]{16}/],
  ["Private key block", /-----BEGIN (?:RSA|DSA|EC|OPENSSH|PGP) PRIVATE KEY-----/],
];

const ASSIGNMENT_PATTERN =
  /^\s*(api[_-]?key|access[_-]?key|app[_-]?key|secret|secret[_-]?key|client[_-]?secret|token|password|passwd|pwd)\s*(?:[:=]|=>)\s*(?:(["'])(.*?)\2|([A-Za-z0-9_./:+-]{8,}))\s*$/i;
const USER_ARRAY_PATTERN = /^\s*(hotwords|recent_context)\s*=\s*\[(.*)$/i;
const CONTEXT_BLOCK_PATTERN = /^\s*\[\[context\.(?:prompt_context|recent_context)\]\]\s*$/i;
const IMAGE_URL_PATTERN = /^\s*image_url\s*=\s*(["'])(.*?)\1\s*$/i;

function runGit(args, cwd) {
  const result = spawnSync("git", args, { cwd, encoding: "utf8" });
  if (result.status !== 0) return null;
  return result.stdout.trim();
}

function repoRoot() {
  const root = runGit(["rev-parse", "--show-toplevel"], process.cwd());
  return root ? path.resolve(root) : null;
}

function normalize(relativePath) {
  return relativePath.split(path.sep).join("/").toLowerCase();
}

function isProtectedLocalFile(relativePath) {
  const normalized = normalize(relativePath);
  const name = path.basename(normalized);
  if (PROTECTED_LOCAL_FILES.has(name)) return true;
  if (name.endsWith(".local.toml")) return true;
  if (normalized.endsWith("/config.toml")) return true;
  if (normalized.endsWith("/voice_input_stats.jsonl")) return true;
  if (name === ".env") return true;
  if (name.startsWith(".env.") && ![".env.example", ".env.sample", ".env.template"].includes(name)) {
    return true;
  }
  return false;
}

function isProbablyBinary(filePath) {
  if (BINARY_SUFFIXES.has(path.extname(filePath).toLowerCase())) return true;
  try {
    const chunk = fs.readFileSync(filePath).subarray(0, 2048);
    return chunk.includes(0);
  } catch {
    return true;
  }
}

function shouldSkip(relativePath, includeProtected = false) {
  if (!includeProtected && isProtectedLocalFile(relativePath)) return true;
  const parts = normalize(relativePath).split("/");
  if (parts.some((part) => EXCLUDE_DIRS.has(part))) return true;
  return isProbablyBinary(path.resolve(relativePath));
}

function looksLikePlaceholder(value) {
  const normalized = value.trim().replace(/^['"]|['"]$/g, "").toLowerCase();
  if (PLACEHOLDER_VALUES.has(normalized)) return true;
  if (/^\d+$/.test(normalized)) return true;
  if (["true", "false"].includes(normalized)) return true;
  return ["example", "placeholder", "示例", "填写", "这里填"].some((word) => normalized.includes(word));
}

function maskSecret(value) {
  if (value.length <= 8) return "*".repeat(value.length);
  return `${value.slice(0, 4)}...${value.slice(-4)}`;
}

function scanFile(filePath) {
  const findings = [];
  let text;
  try {
    text = fs.readFileSync(filePath, "utf8");
  } catch {
    return findings;
  }

  const scanUserContent = [".toml", ".md", ".markdown"].includes(path.extname(filePath).toLowerCase());
  let userArrayKey = null;
  let userArrayReported = false;

  text.split(/\r?\n/).forEach((line, index) => {
    const lineNo = index + 1;
    const stripped = line.trim();

    if (scanUserContent && userArrayKey) {
      if (stripped.startsWith("]")) {
        userArrayKey = null;
        userArrayReported = false;
        return;
      }
      if (!stripped || stripped.startsWith("#") || stripped.startsWith("//")) return;
      if (!userArrayReported) {
        findings.push([lineNo, `Custom ${userArrayKey}`, "non-empty array"]);
        userArrayReported = true;
      }
      return;
    }

    if (!stripped || stripped.startsWith("#") || stripped.startsWith("//")) return;
    if ((stripped.match(/\|/g) || []).length >= 2) return;

    if (scanUserContent && CONTEXT_BLOCK_PATTERN.test(stripped)) {
      findings.push([lineNo, "Custom context block", "context.*"]);
      return;
    }

    const userArray = scanUserContent ? stripped.match(USER_ARRAY_PATTERN) : null;
    if (userArray) {
      const key = userArray[1];
      let value = userArray[2].split("#", 1)[0].trim();
      const closedInline = value.includes("]");
      value = value.split("]", 1)[0].trim();
      if (value) findings.push([lineNo, `Custom ${key}`, "non-empty array"]);
      else if (!closedInline) {
        userArrayKey = key;
        userArrayReported = false;
      }
      return;
    }

    const imageUrl = scanUserContent ? stripped.match(IMAGE_URL_PATTERN) : null;
    if (imageUrl) {
      const value = imageUrl[2].trim();
      if (value && !looksLikePlaceholder(value)) findings.push([lineNo, "Custom image_url", maskSecret(value)]);
      return;
    }

    for (const [name, pattern] of SECRET_PATTERNS) {
      const match = line.match(pattern);
      if (match) findings.push([lineNo, name, maskSecret(match[0])]);
    }

    const assignment = stripped.match(ASSIGNMENT_PATTERN);
    if (assignment) {
      const key = assignment[1];
      const value = (assignment[3] || assignment[4] || "").trim();
      if (value.length >= 8 && !looksLikePlaceholder(value) && !/^[A-Za-z_][A-Za-z0-9_]*$/.test(value)) {
        findings.push([lineNo, `Suspicious ${key}`, maskSecret(value)]);
      }
    }
  });

  return findings;
}

function walkFiles(root) {
  const files = [];
  for (const entry of fs.readdirSync(root, { withFileTypes: true })) {
    const fullPath = path.join(root, entry.name);
    const relative = path.relative(root, fullPath);
    if (entry.isDirectory()) {
      if (!EXCLUDE_DIRS.has(entry.name)) files.push(...walkFiles(fullPath));
    } else if (!shouldSkip(relative)) {
      files.push(fullPath);
    }
  }
  return files;
}

function filesFromGit(root, mode) {
  const output =
    mode === "staged"
      ? runGit(["diff", "--cached", "--name-only", "--diff-filter=ACMR"], root)
      : runGit(["ls-files", "--cached", "--others", "--exclude-standard"], root);
  if (output === null) return null;
  return output
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean)
    .map((relative) => path.join(root, relative))
    .filter((filePath) => fs.existsSync(filePath) && fs.statSync(filePath).isFile());
}

function main() {
  const args = process.argv.slice(2);
  const staged = args.includes("--staged");
  const gitVisible = args.includes("--git-visible");
  const positional = args.filter((arg) => !arg.startsWith("--"));
  if (staged && gitVisible) {
    console.error("[scan-secrets] choose only one of --staged or --git-visible");
    process.exit(2);
  }

  let root = process.cwd();
  let files = [];
  const protectedFilesAreErrors = staged || gitVisible;
  if (staged || gitVisible) {
    const foundRoot = repoRoot();
    if (!foundRoot) {
      console.error("[scan-secrets] not inside a git repository");
      process.exit(2);
    }
    root = foundRoot;
    files = filesFromGit(root, staged ? "staged" : "git-visible");
    if (!files) {
      console.error("[scan-secrets] failed to query git files");
      process.exit(2);
    }
    if (staged && files.length === 0) {
      console.log("[scan-secrets] no staged files to scan");
      return;
    }
  } else {
    const target = path.resolve(positional[0] || ".");
    if (!fs.existsSync(target)) {
      console.error(`[scan-secrets] path not found: ${target}`);
      process.exit(2);
    }
    root = fs.statSync(target).isDirectory() ? target : path.dirname(target);
    files = fs.statSync(target).isDirectory() ? walkFiles(target) : [target];
  }

  let findingsCount = 0;
  for (const filePath of files) {
    const relative = path.relative(root, filePath);
    if (protectedFilesAreErrors && isProtectedLocalFile(relative)) {
      findingsCount += 1;
      console.log(`${normalize(relative)}:0: Protected local file: do not commit local config/log/state`);
      continue;
    }
    if (shouldSkip(relative, protectedFilesAreErrors)) continue;
    for (const [lineNo, kind, masked] of scanFile(filePath)) {
      findingsCount += 1;
      console.log(`${normalize(relative)}:${lineNo}: ${kind}: ${masked}`);
    }
  }

  if (findingsCount) {
    console.log(`[scan-secrets] found ${findingsCount} potential secret(s)`);
    process.exit(1);
  }
  console.log("[scan-secrets] no obvious secrets found");
}

main();
